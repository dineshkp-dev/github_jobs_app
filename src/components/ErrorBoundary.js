import React, { Component } from 'react';
import '../styles/error.css';
import { HOME_PAGE_LINK } from '../common/constants';

export class ErrorBoundary extends Component {
  state = { error: false };

  static getDerivedStateFromError(error) {
    return { error };
  }

  render() {
    const errorMessage =
      this.state.error && this.state.error.message
        ? this.state.error.message
        : 'Oops! An Error has occurred.';
    return this.state.error ? (
      <div className='error-container'>
        <h1 className='error-title'>Error!</h1>
        <div className='error-msg'>
          <div className='error-text'>{errorMessage}</div>
          <div>
            Please click <a href={HOME_PAGE_LINK}>here</a> to go back to the
            home page
          </div>
        </div>
      </div>
    ) : (
      this.props.children
    );
  }
}

export default ErrorBoundary;
