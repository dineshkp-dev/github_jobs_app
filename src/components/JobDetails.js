import React, { useEffect, useState } from 'react';
import '../styles/jobdetails.css';
import parse from 'html-react-parser';
import { getJobDetails } from '../services/jobDetailService';
import { useParams } from 'react-router';
import LoadingIndicator from './LoadingIndicator';
import { useAsyncError } from '../common/commonUtils';
import ListError from './ListError';
import { useHistory } from 'react-router-dom';

function JobDetails() {
  const [jobDetails, setJobDetails] = useState({});
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');

  let { jobid } = useParams();

  console.log(jobid);

  const history = useHistory();
  const throwError = useAsyncError();
  useEffect(() => {
    async function getJobDetailFromService(jobid) {
      let jobDetailResponse = {};
      try {
        jobDetailResponse = await getJobDetails(jobid);
        if (jobDetailResponse.error) {
          setError(true);
          setErrorMsg(jobDetailResponse.errorMsg);
        } else {
          console.log(jobDetailResponse.jobDetail);
          setJobDetails(jobDetailResponse.jobDetail);
        }
      } catch (e) {
        console.log(e);
        throwError(new Error(e));
      }
    }
    getJobDetailFromService(jobid);
  }, [throwError, jobid]);

  return (
    <React.Fragment>
      {Object.keys(jobDetails).length > 0 &&
      jobDetails.constructor === Object ? (
        <React.Fragment>
          <div className='jobdetails-header-container'>
            <div className='jobdetails-goback' onClick={() => history.goBack()}>
              <i class='fas fa-chevron-left'></i> Go Back
            </div>
            <div className='jobdetails-jobtype'>Job type: {jobDetails.type}</div>
            <div className='jobdetails-location'>Job location: {jobDetails.location}</div>
            <div className='jobdetails-createdat'>Created at: {jobDetails.created_at}</div>
            <div className='jobdetails-title'>Job title: {jobDetails.title}</div>
          </div>
          <hr />
          <div className='jobdetails-companydetails-container'>
          <div className='jobdetails-sub-section'>Company Details</div>
            <div className='jobdetails-company'>{jobDetails.company}</div>
            <div className='jobdetails-companyurl'>
              {jobDetails.company_url}
            </div>
          </div>
          <hr />
          <div className='jobdetails-descripton-container'>
            <div className='jobdetails-sub-section'>Job Details</div>
            <div className='jobdetails-description'>
              {parse('' + jobDetails.description)}
            </div>
          </div>
          <hr />
          <div className='jobdetails-howtoapply-container'>
            <div className='jobdetails-sub-section'>How To Apply</div>
            <div className='jobdetails-howtoapply'>
              {parse('' + jobDetails.how_to_apply)}
            </div>
          </div>
        </React.Fragment>
      ) : error ? (
        <ListError text={errorMsg} />
      ) : (
        <LoadingIndicator text='Please wait for job details to be updated' />
      )}
    </React.Fragment>
  );
}

export default JobDetails;
