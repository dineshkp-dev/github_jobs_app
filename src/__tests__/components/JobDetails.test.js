import JobDetails from '../../components/JobDetails';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { jobdetails } from '../utils/testUtils';
import ErrorBoundary from '../../components/ErrorBoundary';
import '@testing-library/jest-dom/extend-expect';

const jobDetailService = require('../../services/jobDetailService');

describe('JobDetails snapshot test', () => {
  test('renders', () => {
    jobDetailService.getJobDetails = jest.fn(() => Promise.resolve(jobdetails));
    const { container } = render(
      <MemoryRouter initialEntries={['/testid']}>
        <JobDetails />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});

describe('JobDetails render test', () => {
  test('renders JobDetails', () => {
    jobDetailService.getJobDetails = jest.fn(() => Promise.resolve(jobdetails));
    render(
      <MemoryRouter initialEntries={['/testid']}>
        <JobDetails />
      </MemoryRouter>
    );
    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(
      screen.getByText('Please wait for job details to be updated')
    ).toBeInTheDocument();
  });
});

describe('JobDetails get results test', () => {
  test('get success results from Details Service', async () => {
    jobDetailService.getJobDetails = jest.fn(() => Promise.resolve(jobdetails));
    render(
      <MemoryRouter initialEntries={['/testid']}>
        <JobDetails />
      </MemoryRouter>
    );
    await waitFor(() => {
      expect(screen.getByText('Job type: Full Time')).toBeInTheDocument();
      expect(screen.getByText('Job Details')).toBeInTheDocument();
      expect(
        screen.getByText('Job title: Full Stack Engineer')
      ).toBeInTheDocument();
      expect(
        screen.getByText('Job location: New York City')
      ).toBeInTheDocument();
      expect(
        screen.getByText('Please apply at the following website:')
      ).toBeInTheDocument();
    });
  });

  test('Job Details and go back', async () => {
    jobDetailService.getJobDetails = jest.fn(() => Promise.resolve(jobdetails));
    render(
      <MemoryRouter initialEntries={['/testid']}>
        <JobDetails />
      </MemoryRouter>
    );
    await waitFor(() => {
      expect(screen.getByText('Job type: Full Time')).toBeInTheDocument();
      expect(screen.getByText('Job Details')).toBeInTheDocument();
      expect(
        screen.getByText('Job title: Full Stack Engineer')
      ).toBeInTheDocument();
      expect(
        screen.getByText('Job location: New York City')
      ).toBeInTheDocument();
      expect(
        screen.getByText('Please apply at the following website:')
      ).toBeInTheDocument();
    });
    const goBack = screen.getByText(/Go Back/);
    fireEvent.click(goBack);
    await waitFor(() => {
      expect(screen.getByText('Job Details')).toBeInTheDocument();
    });
  });
});

describe('get empty results from JobDetails service test', () => {
  test('get dummy errorf results from Details Service', async () => {
    let errorListings = jobdetails;
    errorListings.error = true;
    errorListings.errorMsg = 'Dummy error message';
    jobDetailService.getJobDetails = jest.fn(() =>
      Promise.resolve(errorListings)
    );
    render(
      <MemoryRouter initialEntries={['/testid']}>
        <JobDetails />
      </MemoryRouter>
    );
    await waitFor(() => {
      expect(screen.getByText('No listing found')).toBeInTheDocument();
      expect(screen.getByText('Dummy error message')).toBeInTheDocument();
    });
  });

  test('Reject from JobDetails Service', async () => {
    jobDetailService.getJobDetails = jest.fn(() =>
      Promise.reject('DUMMY error')
    );
    render(
      <MemoryRouter initialEntries={['/testid']}>
        <ErrorBoundary>
          <JobDetails />
        </ErrorBoundary>
      </MemoryRouter>
    );
    await waitFor(() => {
      expect(screen.getByText('Error!')).toBeInTheDocument();
      expect(screen.getByText('DUMMY error')).toBeInTheDocument();
      expect(screen.getByText(/Please click/i)).toBeInTheDocument();
    });
  });

  test('Exception from JobDetails Service', async () => {
    jobDetailService.getJobDetails = jest.fn(() => {
      throw new Error('DUMMY error');
    });
    render(
      <MemoryRouter initialEntries={['/testid']}>
        <ErrorBoundary>
          <JobDetails />
        </ErrorBoundary>
      </MemoryRouter>
    );
    await waitFor(() => {
      expect(screen.getByText('Error!')).toBeInTheDocument();
      expect(screen.getByText('Error: DUMMY error')).toBeInTheDocument();
      expect(screen.getByText(/Please click/i)).toBeInTheDocument();
    });
  });
});
