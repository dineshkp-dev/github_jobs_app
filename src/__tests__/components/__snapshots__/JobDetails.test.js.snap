// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobDetails snapshot test renders 1`] = `
<div>
  <div
    class="loading-indicator-container"
  >
    <h3
      class="loading-title"
    >
      Loading...
    </h3>
    <div
      class="loading-text"
    >
      Please wait for job details to be updated
    </div>
  </div>
</div>
`;
